src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -61,20 +72,158 @@ private static void CheckIsLastArgumentCancellationTokenOrRequestContext(ISymbol
                     context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0002, member.Locations.FirstOrDefault()), member);
                 }
             }
-            else if (IsCancellationToken(lastArgument) && !lastArgument.IsOptional)
+            else if (IsCancellationToken(lastArgument))
             {
-                var overloadWithCancellationToken = FindMethod(
+                if (!lastArgument.IsOptional)
+                {
+                    var overloadWithCancellationToken = FindMethod(
                     member.ContainingType.GetMembers(member.Name).OfType<IMethodSymbol>(),
                     member.TypeParameters,
                     member.Parameters.RemoveAt(member.Parameters.Length - 1));
 
-                if (overloadWithCancellationToken == null)
+                    if (overloadWithCancellationToken == null)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0002, member.Locations.FirstOrDefault()), member);
+                    }
+                }
+
+                if (member.Parameters.FirstOrDefault(IsRequestContent) != null)
                 {
-                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0002, member.Locations.FirstOrDefault()), member);
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0017, member.Locations.FirstOrDefault()), member);
+                }
+            }
+            else if (IsRequestContext(lastArgument))
+            {
+                CheckProtocolMethodReturnType(context, member);
+                CheckProtocolMethodParameters(context, member);
+            }
+        }
+
+        private static string GetFullNamespaceName(IParameterSymbol parameter)
+        {
+            var currentNamespace = parameter.Type.ContainingNamespace;
+            string currentName = currentNamespace.Name;
+            string fullNamespace = "";
+            while (!string.IsNullOrEmpty(currentName))
+            {
+                fullNamespace = fullNamespace == "" ? currentName : $"{currentName}.{fullNamespace}";
+                currentNamespace = currentNamespace.ContainingNamespace;
+                currentName = currentNamespace.Name;
+            }
+            return fullNamespace;
+        }
+
+        // A protocol method should not have model as parameter. If it has ambiguity with convenience method, it should have required RequestContext.
+        // Ambiguity: doesn't have a RequestContent, but there is a method ending with CancellationToken has same type of parameters 
+        // No ambiguity: has RequestContent.
+        private static void CheckProtocolMethodParameters(ISymbolAnalysisContext context, IMethodSymbol method)
```

Comment:
would a better name for this method be something like - `CheckforAmbiguousMethodPair` ?

Reply:
But this method not only checks the ambiguity, but also the input parameters and output parameters of protocol methods.